Populate when we send an email request to Notify

1pretz1 opened the pull request:
This PR is the first stage of work to do with reducing our reliance on status updates from notify to archive emails. 

Included in this PR:

* Created a new `sent_at` column on Email model
* Updated the `sending` enum to `sent` for the DeliveryAttempt model
* We now populate the new column (sent_at) and mark the email as sent when we successfully make an email request to Notify. 
* Stop updating the email's status from the status update

Future work to do:

* Remove sent_at and completed_at from the DeliveryAttempt model
* Remove finished_sending_at from the Email model
* Delete all emails older than 14 days old

Trello:
https://trello.com/c/15eVh9Qg/482-stop-relying-on-notify-status-updates-to-archive-delete-emails

benthorner commented on app/services/delivery_request_service.rb:
```diff
@@ -31,13 +31,21 @@ def call
+      sent_at = attempt.finished_sending_at
+
+      if attempt.delivered?
+        email.update!(status: :sent, finished_sending_at: sent_at, sent_at: sent_at)
```
```suggestion
        email.update!(status: :sent, finished_sending_at: sent_at, sent_at: Time.zone.now)
```

We can be consistent with the line above here

benthorner commented on app/services/delivery_request_service.rb:
```diff
@@ -31,13 +31,21 @@ def call
     status = Metrics.email_send_request(provider_name) { send_email }
 
-    return attempt if status == :sent
+    if status == :sent
```
Could we simplify this code a bit? It's confusing that we're:

- Doing some of the updates in a transaction.
- Sometimes checking `attempt` vs. `status`.
- Implicitly handling the other (comms) status.

We could have something like:

```
ActiveRecord::Base.transaction do
  case status
  when :sent ...
  when :delivered ...
  when :undeliverable_failure ...
  when :provider_communication_failure ...
end
```

benthorner replied:
Although this may introduce a little duplication, I think it's worth it for the clarity.

1pretz1 replied:
Sure thing, I've rejigged this bit around and added some missing tests around recording metrics. I did actually start writing a case statement but then changed my mind, have added said case.

benthorner commented on app/services/status_update_service.rb:
```diff
@@ -70,9 +70,9 @@ def delivery_attempt_status
-    finished_sending_at = delivery_attempt.finished_sending_at
-    email.mark_as_sent(finished_sending_at) if delivery_attempt.delivered?
-    email.mark_as_failed(finished_sending_at) if delivery_attempt.undeliverable_failure?
+    if delivery_attempt.status.in? %w[delivered undeliverable_failure]
```
Do we need this `if` anymore?

1pretz1 replied:
Nope, good spot!

benthorner commented on app/services/delivery_request_service.rb:
```diff
@@ -31,13 +31,21 @@ def call
+      if attempt.delivered?
+        email.update!(status: :sent, finished_sending_at: sent_at, sent_at: sent_at)
+      elsif attempt.undeliverable_failure?
+        email.update!(status: :failed, finished_sending_at: sent_at, sent_at: sent_at)
```
We shouldn't populate `sent_at` if we haven't sent the email successfully.

1pretz1 replied:
Ah for some reason I thought only for a communication failure we shouldn't update sent_at, got my wires crossed!

benthorner commented on spec/models/delivery_attempt_spec.rb:
```diff
@@ -2,7 +2,7 @@
     subject { delivery_attempt.finished_sending_at }
 
-    context "when email is sent" do
+    context "when email is delivered" do
```
Nice catch!

benthorner reviewed (commented):
Thanks for breaking this into separate commits 👍. I found an issue with populating `sent_at` in the last commit, and also have a suggestion to make the code a bit easier to understand, as I struggled a bit.

Since this PR is something of a halfway point for the next two weeks, I also think it's worth adding a "DEPRECATED" comment in places where we populate `finished_sending_at`. The reason is that this field no longer makes much sense: we can be "finished sending" but have `sent_at: nil`!

1pretz1 commented:
Thanks for the review @benthorner, I've addressed your comments and added come deprecation comments.

It also looks like the CI hooks are pointing to Carrenza again so we have passing tests! 🎉

benthorner commented on spec/services/delivery_request_service_spec.rb:
```diff
@@ -155,10 +168,18 @@
+      it "marks the email as failed and sets sent_at time" do
+        freeze_time do
+          expect { described_class.call(email: email) }
+            .to change { email.status }.to("failed")
```
We're missing the actual check for `sent_at` here.

1pretz1 replied:
That was old text describing the test - have updated it accordingly!

benthorner commented on spec/services/status_update_service_spec.rb:
```diff
@@ -26,18 +26,13 @@
-          .to change { delivery_attempt.reload.status }
-          .to("delivered")
+          .to change { delivery_attempt.reload.status }.to("delivered")
+          .and change { email.reload.finished_sending_at }.to(sent_at)
```
Minor: the use of `sent_at` is (now) confusing here, and seems redundant since we also have this `completed_at` variable, which is set to the same value. Can we just say `timestamp` or `status_update_time` (or something)?

benthorner reviewed (approved):
Thanks for adding the deprecation warnings @1pretz1 - looks good! I found a couple of minor issues that would be good to fix before this goes in.
